src/librustc_traits/lowering.rs

```diff
@@ -283,9 +282,7 @@ fn program_clauses_for_trait<'a, 'tcx>(
 
     // `FromEnv(WC) :- FromEnv(Self: Trait<P1..Pn>)`, for each where clause WC
     let implied_bound_clauses = where_clauses
-        .iter()
-        .cloned()
-
+        .clone()
```

Comment:
Hm, isn't this cloning the underlying vector?

Reply:
I thought this only clones the `Iter` struct, i.e. only the start and end pointer (and the marker)?

Reply:
`where_clause` is an into_iter struct, not an iter struct which is why I think this will copy the vector.

Reply:
Ah, you're right; luckily, there is still room for improvement - I'll update the PR shortly.